File: internal/task/task.go
package task

import (
	"fmt"
	"os"

	"gopkg.in/yaml.v3"
)

// Task represents a task definition loaded from YAML
type Task struct {
	ID              string            `yaml:"id"`
	Name            string            `yaml:"name"`
	Type            string            `yaml:"type"`
	Repo            string            `yaml:"repo"`
	StartingCommit  string            `yaml:"starting_commit"`
	Prompt          string            `yaml:"prompt"`
	Prompts         []string          `yaml:"prompts"` // for type: sequential
	SuccessCriteria []Criterion       `yaml:"success_criteria"`
	QualityRubric   []RubricDimension `yaml:"quality_rubric"`
}

// Criterion represents a success criterion with typed fields
type Criterion struct {
	Type     string   `yaml:"type"`
	Paths    []string `yaml:"paths"`
	Command  string   `yaml:"command"`
	Contains []string `yaml:"contains"`
}

// RubricDimension represents a quality rubric dimension
type RubricDimension struct {
	Dimension string `yaml:"dimension"`
	Prompt    string `yaml:"prompt"`
}

// LoadTask loads a task definition from a YAML file
func LoadTask(filename string) (*Task, error) {
	data, err := os.ReadFile(filename)
	if err != nil {
		return nil, fmt.Errorf("failed to read file %s: %w", filename, err)
	}

	var task Task
	if err := yaml.Unmarshal(data, &task); err != nil {
		return nil, fmt.Errorf("failed to parse YAML: %w", err)
	}

	// Set default type if not specified
	if task.Type == "" {
		if len(task.Prompts) > 0 {
			task.Type = "sequential"
		} else {
			task.Type = "single_shot"
		}
	}

	// Validate required fields
	if task.ID == "" {
		return nil, fmt.Errorf("task id is required")
	}

	switch task.Type {
	case "sequential":
		if len(task.Prompts) < 2 {
			return nil, fmt.Errorf("sequential task requires at least 2 prompts")
		}
	default:
		if task.Prompt == "" {
			return nil, fmt.Errorf("prompt is required for single_shot tasks")
		}
	}

	return &task, nil
}

File: src/token_miser/task.py
"""Task YAML loading and validation."""
from __future__ import annotations

import os
from dataclasses import dataclass, field
from pathlib import Path

import yaml


@dataclass
class Criterion:
    type: str
    paths: list[str] = field(default_factory=list)
    command: str = ""
    contains: list[str] = field(default_factory=list)


@dataclass
class RubricDimension:
    dimension: str
    prompt: str


@dataclass
class Task:
    id: str
    name: str
    repo: str
    starting_commit: str
    prompt: str = ""
    prompts: list[str] = field(default_factory=list)
    type: str = "single_shot"
    success_criteria: list[Criterion] = field(default_factory=list)
    quality_rubric: list[RubricDimension] = field(default_factory=list)
    repo_id: str = ""
    category: str = ""
    setup_commands: list[str] = field(default_factory=list)


def load_task(filename: str | Path) -> Task:
    """Load and validate a task YAML file."""
    path = Path(filename)
    data = yaml.safe_load(path.read_text())

    task_id = data.get("id", "")
    if not task_id:
        raise ValueError(f"Task file {filename} missing required 'id' field")

    prompts = data.get("prompts", [])
    prompt = data.get("prompt", "")
    task_type = data.get("type", "")

    if not task_type:
        task_type = "sequential" if prompts else "single_shot"

    if task_type == "sequential" and len(prompts) < 2:
        raise ValueError(f"Sequential task {task_id} requires at least 2 prompts")
    if task_type == "single_shot" and not prompt:
        raise ValueError(f"Single-shot task {task_id} requires a 'prompt' field")

    criteria = [
        Criterion(
            type=c.get("type", ""),
            paths=c.get("paths", []),
            command=c.get("command", ""),
            contains=c.get("contains", []),
        )
        for c in data.get("success_criteria", [])
    ]

    rubric = [
        RubricDimension(dimension=r["dimension"], prompt=r["prompt"])
        for r in data.get("quality_rubric", [])
    ]

    # Resolve ${VAR} references in repo path from environment
    repo = data.get("repo", "")
    repo = os.path.expandvars(repo)

    return Task(
        id=task_id,
        name=data.get("name", ""),
        repo=repo,
        starting_commit=data.get("starting_commit", ""),
        prompt=prompt,
        prompts=prompts,
        type=task_type,
        success_criteria=criteria,
        quality_rubric=rubric,
        repo_id=data.get("repo_id", ""),
        category=data.get("category", ""),
        setup_commands=data.get("setup_commands", []),
    )
